# Proyect 7 using Streamlit 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import math
from scipy import stats as st

In [2]:
# Carga de archivos de DataFrame

print("#############################################")
print("#                VEHICLES-US                #")
print("#############################################")

print("\n╔════════════════════════════════╗")  
print("║   ANÁLISIS DE CIENCIA DE DATOS ║")  
print("║         DESARROLLADO POR       ║")  
print("║    Mauricio Zárate Romero      ║")  
print("╚════════════════════════════════╝") 

# car_data = df_cd

df_cd_copy = pd.read_csv('vehicles_us.csv')

# Copia de seguridad para no modificar los DataFrames originales.

df_cd = df_cd_copy.copy()

#############################################
#                VEHICLES-US                #
#############################################

╔════════════════════════════════╗
║   ANÁLISIS DE CIENCIA DE DATOS ║
║         DESARROLLADO POR       ║
║    Mauricio Zárate Romero      ║
╚════════════════════════════════╝


In [3]:
# Información general/resumida sobre el DataFrame

print("\n╔══════════════════════════════════════════════════╗")
print("║                'vehicles_us.csv'                 ║")
print("║                  'Información'                   ║")
print("╚══════════════════════════════════════════════════╝")
print("\n" + "=" * 50)
df_cd.info()
print("\n" + "=" * 50)
print("\n- Total de filas duplicadas:", df_cd.duplicated().sum())
print("\n" + "=" * 50)
print('\n- Total de valores nulos:\n \n', df_cd.isna().sum())
print("\n" + "=" * 50)
print('\n- Estadisticas\n \n', df_cd.describe())


╔══════════════════════════════════════════════════╗
║                'vehicles_us.csv'                 ║
║                  'Información'                   ║
╚══════════════════════════════════════════════════╝

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), ob

In [4]:
df_cd.sample(10)

price  model_year                  model  condition  cylinders    fuel  \
9335    5950      2012.0          nissan altima       good        4.0     gas   
14230   6900      2011.0         hyundai sonata  excellent        NaN     gas   
21921   3695      2010.0           nissan versa  excellent        4.0     gas   
2853   26879         NaN    jeep grand cherokee       good        6.0     gas   
51425   8495      2006.0             ford f-150       good        8.0     gas   
22417  27988      2015.0    chevrolet silverado   like new        8.0     gas   
29559  10878      2015.0       chevrolet impala  excellent        6.0     gas   
5702   34990      2019.0             ford f-150  excellent        8.0     gas   
43348  11800      2005.0  ford f-250 super duty       good        8.0  diesel   
39747  13998      2016.0      chevrolet equinox   like new        4.0     gas   

       odometer transmission    type paint_color  is_4wd date_posted  \
9335   134428.0    automatic   sedan        blue     NaN  2019-03-10   
14230  106252.0    automatic   sedan        grey     NaN  2018-06-25   
21921   74000.0       manual   sedan         red     NaN  2019-03-30   
2853    29991.0    automatic     SUV       black     1.0  2019-03-23   
51425   96926.0    automatic   truck         NaN     1.0  2018-09-23   
22417   35000.0    automatic   truck         NaN     NaN  2018-07-30   
29559  106127.0    automatic   sedan         red     NaN  2018-08-21   
5702        NaN    automatic   truck       white     1.0  2019-01-08   
43348  169000.0    automatic  pickup       black     1.0  2018-10-26   
39747   60000.0    automatic     SUV        blue     NaN  2019-03-12   

       days_listed  
9335            80  
14230           17  
21921           31  
2853            25  
51425           28  
22417           36  
29559           38  
5702            22  
43348            8  
39747           17

### Tratamiento por columna

### price

In [5]:
# 1. Mostrar valores nulos y tipo de columna antes de modificar
print(f"Valores nulos = {df_cd['price'].isna().sum()}")
print(f"Tipo de columna = {df_cd['price'].dtypes}")

# 2. Reemplazar los valores 1 por NaN
indices_1 = df_cd[df_cd['price'] == 1].index.tolist()
df_cd.loc[indices_1, 'price'] = np.nan

# 3. Mostrar valores nulos y tipo de columna después de modificar
print(f"Valores nulos = {df_cd['price'].isna().sum()}")
print(f"Tipo de columna = {df_cd['price'].dtypes}")

# 4. Comprobar que ya no hay valores 1 en 'price'
indices_1 = df_cd[df_cd['price'] == 1].index.tolist()
print(f"Índices con price = 1: {indices_1}")

# 5. Eliminar filas con NaN en 'price' (opcional, si quieres limpiar el DataFrame)
df_cd = df_cd.dropna(subset=['price'])

# 6. Mostrar valores nulos tras eliminar filas
print(f"Valores nulos después de dropna = {df_cd['price'].isna().sum()}")

# 7. Cambiar la columna a int
df_cd['price'] = df_cd['price'].astype(int)
print(df_cd['price'].dtype)


Valores nulos = 0
Tipo de columna = int64
Valores nulos = 798
Tipo de columna = float64
Índices con price = 1: []
Valores nulos después de dropna = 0
int64


### model_year

In [6]:
# Mostrar estadísticas descriptivas de la columna 'model_year' antes de eliminar nulos
print(df_cd['model_year'].describe())

# Imprimir la cantidad de valores nulos en la columna 'model_year' antes de eliminarlos
print(f'\nvalores nulos = {df_cd["model_year"].isna().sum()}')

# Eliminar filas donde 'model_year' es nulo
df_cd = df_cd.dropna(subset=['model_year'])

# Imprimir la cantidad de valores nulos en la columna 'model_year' después de eliminarlos
print(f'\nvalores nulos despues de limpiarlos = {df_cd["model_year"].isna().sum()}')

df_cd['model_year'] = df_cd['model_year'].astype(int)

print(f'\nCambio a tipo de dato = {df_cd['model_year'].dtype}')


valores_ordenados_my = df_cd['model_year'].drop_duplicates().sort_values()

valores_ordenados_my = np.unique(df_cd['model_year'])
print(valores_ordenados_my)









count    47160.000000
mean      2009.657591
std          6.262325
min       1908.000000
25%       2006.000000
50%       2011.000000
75%       2014.000000
max       2019.000000
Name: model_year, dtype: float64

valores nulos = 3567

valores nulos despues de limpiarlos = 0

Cambio a tipo de dato = int64
[1908 1929 1936 1948 1949 1954 1955 1958 1960 1961 1962 1963 1964 1965
 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978 1979
 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993
 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007
 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019]


### model

In [7]:
print(df_cd['model'].describe())

print(f'\nvalores nulos = {df_cd["model"].isna().sum()}')

valores_ordenados_m = np.sort(np.unique(df_cd['model']))
print('\n')
print(valores_ordenados_m)

print(f'\nEl tipo de columna es correcto = {df_cd['model'].dtype}')

count          47160
unique           100
top       ford f-150
freq            2548
Name: model, dtype: object

valores nulos = 0


['acura tl' 'bmw x5' 'buick enclave' 'cadillac escalade'
 'chevrolet camaro' 'chevrolet camaro lt coupe 2d' 'chevrolet colorado'
 'chevrolet corvette' 'chevrolet cruze' 'chevrolet equinox'
 'chevrolet impala' 'chevrolet malibu' 'chevrolet silverado'
 'chevrolet silverado 1500' 'chevrolet silverado 1500 crew'
 'chevrolet silverado 2500hd' 'chevrolet silverado 3500hd'
 'chevrolet suburban' 'chevrolet tahoe' 'chevrolet trailblazer'
 'chevrolet traverse' 'chrysler 200' 'chrysler 300'
 'chrysler town & country' 'dodge charger' 'dodge dakota'
 'dodge grand caravan' 'ford econoline' 'ford edge' 'ford escape'
 'ford expedition' 'ford explorer' 'ford f-150' 'ford f-250'
 'ford f-250 sd' 'ford f-250 super duty' 'ford f-350 sd' 'ford f150'
 'ford f150 supercrew cab xlt' 'ford f250' 'ford f250 super duty'
 'ford f350' 'ford f350 super duty' 'ford focus' 'ford focus se

Al parecer la columna no necesita tratamiento

### condition

In [8]:
print(df_cd['condition'].describe())

print(f'\nvalores nulos = {df_cd["condition"].isna().sum()}')

valores_ordenados_c = np.sort(np.unique(df_cd['condition']))
print('\n')
print(valores_ordenados_c)

# Filtrar filas donde 'condition' es 'new' y 'odometer' no es 0
condicion_new_odometro_no_cero = df_cd[(df_cd['condition'] == 'new') & (df_cd['odometer'] != 0)]
print(f"\nCantidad de autos 'new' con odómetro distinto de 0: {len(condicion_new_odometro_no_cero)}")

condicion_new_odometro_si_cero = df_cd[(df_cd['condition'] == 'new') & (df_cd['odometer'] == 0)]
print(f"\nCantidad de autos 'new' con odómetro distinto de 0: {len(condicion_new_odometro_si_cero)}")




print(f'\nEl tipo de columna es correcto = {df_cd['condition'].dtype}')

count         47160
unique            6
top       excellent
freq          22382
Name: condition, dtype: object

valores nulos = 0


['excellent' 'fair' 'good' 'like new' 'new' 'salvage']

Cantidad de autos 'new' con odómetro distinto de 0: 130

Cantidad de autos 'new' con odómetro distinto de 0: 0

El tipo de columna es correcto = object


por lo que parece la columna de condition es erronea por lo que se le dara tratamiento estableciendo nuevos parametros para definir la condicion del vehiculo y quitando la columna de new ya que todos los vehiculos son usados y no deberia de haber datos con odometro en 0

In [9]:
def nuevas_condiciones(odometer):
    if pd.isna(odometer) or odometer == 0:
        return np.nan
    elif odometer <= 10000:
        return 'like new'
    elif odometer <= 40000:
        return 'excellent'
    elif odometer <= 100000:
        return 'good'
    elif odometer <= 180000:
        return 'fair'
    else:
        return 'salvage'
    
df_cd['condition_fixed'] = df_cd['odometer'].apply(nuevas_condiciones)


df_cd = df_cd.dropna(subset='condition_fixed').copy()


df_cd.drop(columns=['condition'], inplace=True)
df_cd.rename(columns={'condition_fixed' : 'condition'}, inplace=True)

print(df_cd['condition'].value_counts())

print('\n')

df_cd.info()

condition
fair         17684
good         11459
salvage       5869
excellent     3618
like new      1130
Name: count, dtype: int64


<class 'pandas.core.frame.DataFrame'>
Index: 39760 entries, 0 to 51523
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         39760 non-null  int64  
 1   model_year    39760 non-null  int64  
 2   model         39760 non-null  object 
 3   cylinders     35690 non-null  float64
 4   fuel          39760 non-null  object 
 5   odometer      39760 non-null  float64
 6   transmission  39760 non-null  object 
 7   type          39760 non-null  object 
 8   paint_color   32649 non-null  object 
 9   is_4wd        19578 non-null  float64
 10  date_posted   39760 non-null  object 
 11  days_listed   39760 non-null  int64  
 12  condition     39760 non-null  object 
dtypes: float64(3), int64(3), object(7)
memory usage: 4.2+ MB


### cylinders

In [10]:
print(df_cd['cylinders'].describe())

print(f'\nvalores nulos = {df_cd["cylinders"].isna().sum()}')

df_cd = df_cd.dropna(subset=['cylinders'])

print(f'\nvalores nulos despues de limpiarlos = {df_cd["cylinders"].isna().sum()}')

df_cd['cylinders'] = df_cd['cylinders'].astype(int)
print(f'\nTipo de dato se cambia de float a int = {df_cd["cylinders"].dtype}')

count    35690.000000
mean         6.105940
std          1.636197
min          3.000000
25%          4.000000
50%          6.000000
75%          8.000000
max         12.000000
Name: cylinders, dtype: float64

valores nulos = 4070

valores nulos despues de limpiarlos = 0

Tipo de dato se cambia de float a int = int64


### fuel

In [11]:
print(df_cd['fuel'].describe())

print(f'\nvalores nulos = {df_cd["fuel"].isna().sum()}')

valores_ordenados_f = sorted(pd.unique(df_cd['fuel']))
print(f'\n{valores_ordenados_f}')

conteo_other = (df_cd['fuel'] == 'other').sum()
print(f'\nEn la columna de fuel se repite "other" {conteo_other} veces')

df_cd = df_cd[df_cd['fuel'] != 'other']

valores_ordenados_f = sorted(pd.unique(df_cd['fuel']))
print(f'\n{valores_ordenados_f}')

conteo_other = (df_cd['fuel'] == 'other').sum()
print(f'\nEn la columna de fuel se repite "other" {conteo_other} veces')

print(f'\nEl tipo de columna es correcto = {df_cd['fuel'].dtype}')

count     35690
unique        5
top         gas
freq      32772
Name: fuel, dtype: object

valores nulos = 0

['diesel', 'electric', 'gas', 'hybrid', 'other']

En la columna de fuel se repite "other" 68 veces

['diesel', 'electric', 'gas', 'hybrid']

En la columna de fuel se repite "other" 0 veces

El tipo de columna es correcto = object


Se quitaron los valores de 'other' ya que representa poco y no da informacion

### odometer

In [12]:
print(df_cd['odometer'].describe())

print(f'\nvalores nulos = {df_cd["odometer"].isna().sum()}')

df_cd = df_cd.dropna(subset='odometer')

print(f'\nvalores nulos despues de limpiarlos = {df_cd["cylinders"].isna().sum()}')

print(f'\nEl tipo de columna es correcto = {df_cd['cylinders'].dtype}')

odometer_cero = df_cd[df_cd['odometer'] == 0].value_counts()

# Imprimir el DataFrame filtrado
print(odometer_cero)

print(f"\nCantidad de vehículos con odómetro en 0: {len(odometer_cero)}")

print(f'\nEl tipo de columna es correcto = {df_cd['odometer'].dtype}')

count     35622.000000
mean     117108.697097
std       64411.726935
min           1.000000
25%       72705.500000
50%      114473.500000
75%      156000.000000
max      990000.000000
Name: odometer, dtype: float64

valores nulos = 0

valores nulos despues de limpiarlos = 0

El tipo de columna es correcto = int64
Series([], Name: count, dtype: int64)

Cantidad de vehículos con odómetro en 0: 0

El tipo de columna es correcto = float64


fue aqui donde se detecto el error entre 'odometer' con valores altos y 'condicion' 'new' por lo que se quitaron los valores nulos desde el tratamiento de la columna 'condition' 

### transmission

In [13]:
print(df_cd['transmission'].describe())

print(f'\nvalores nulos = {df_cd["transmission"].isna().sum()}')

print('\n')
valores_ordenados_t = sorted(pd.unique(df_cd['transmission']))
print(valores_ordenados_t)

conteo_transmission = df_cd['transmission'].value_counts()
print(f'\nEl tipo de "transmission" "other" se repite {conteo_transmission.get('other', 0)} veces')

print(f'\nEl tipo de columna es correcto = {df_cd['transmission'].dtype}')

count         35622
unique            3
top       automatic
freq          32601
Name: transmission, dtype: object

valores nulos = 0


['automatic', 'manual', 'other']

El tipo de "transmission" "other" se repite 1049 veces

El tipo de columna es correcto = object


Tome la decision de dejar el valor 'other' en esta columna ya que si hay mas tipos de transmiciones.

### type

In [14]:
print(df_cd['type'].describe())

print(f'\nvalores nulos = {df_cd["type"].isna().sum()}')

valores_ordenados_ty = sorted(pd.unique(df_cd['type']))

print(f'\nEl tipo de columna es correcto = {df_cd['type'].dtype}')

count     35622
unique       13
top         SUV
freq       8567
Name: type, dtype: object

valores nulos = 0

El tipo de columna es correcto = object


se reviso la columna y todo parece en orden

### paint_color

In [15]:
print(df_cd['paint_color'].describe())

print(f'\nvalores nulos = {df_cd["paint_color"].isna().sum()}')

df_cd['paint_color'] = df_cd['paint_color'].fillna('Unknown')

print(f'\nvalores nulos despues de la limpieza = {df_cd["paint_color"].isna().sum()}')

valores_ordenados_p = sorted(pd.unique(df_cd['paint_color']))

print(f'\nEl tipo de columna es correcto = {df_cd['paint_color'].dtype}')

count     29269
unique       12
top       white
freq       6987
Name: paint_color, dtype: object

valores nulos = 6353

valores nulos despues de la limpieza = 0

El tipo de columna es correcto = object


### is_4wd

In [16]:
print(df_cd['is_4wd'].describe())

print(f'\nvalores nulos = {df_cd["is_4wd"].isna().sum()}')

df_cd['is_4wd'] = df_cd['is_4wd'].fillna(0)

print(f'\nvalores nulos despues de la limpieza = {df_cd["paint_color"].isna().sum()}')

df_cd['is_4wd'] = df_cd['is_4wd'].astype(int)

print(f'\nEl tipo de columna se cambio de float a int = {df_cd['is_4wd'].dtype}')

count    17548.0
mean         1.0
std          0.0
min          1.0
25%          1.0
50%          1.0
75%          1.0
max          1.0
Name: is_4wd, dtype: float64



valores nulos = 18074

valores nulos despues de la limpieza = 0

El tipo de columna se cambio de float a int = int64


la columna tenia 18074 valores nulos pero al parecer es error de captura por lo que se cambio a 0 debido a que el valor capturado eraa de 1 que corresponde a que el vehículo tiene tracción en las cuatro ruedas (4WD).

### date_posted

In [17]:
print(df_cd['date_posted'].describe())

print(f'\nvalores nulos = {df_cd["date_posted"].isna().sum()}')

df_cd['date_posted'] = pd.to_datetime(df_cd['date_posted'], format='%Y-%m-%d')

print(f'\nEl tipo de columna se caambio de object a date correcto = {df_cd['date_posted'].dtype}')

count          35622
unique           354
top       2019-03-17
freq             128
Name: date_posted, dtype: object

valores nulos = 0

El tipo de columna se caambio de object a date correcto = datetime64[ns]


### days_listed

In [18]:
print(df_cd['days_listed'].describe())

print(f'\nvalores nulos = {df_cd["days_listed"].isna().sum()}')

valores_ordenados_d = sorted(pd.unique(df_cd['days_listed']))

dias_cero = df_cd[df_cd['days_listed'] == 0]

print(len(dias_cero))

dias_cero.sample(10)

count    35622.000000
mean        39.686289
std         28.132268
min          0.000000
25%         19.000000
50%         33.000000
75%         54.000000
max        271.000000
Name: days_listed, dtype: float64

valores nulos = 0
37


price  model_year                     model  cylinders fuel  odometer  \
41940   1800        2002               honda civic          4  gas  238155.0   
43590   2800        2003              ford mustang          6  gas  181000.0   
35605   1500        1992              honda accord          4  gas  239199.0   
26876   8950        2005  chevrolet silverado 1500          8  gas  175788.0   
23195   7995        2013            toyota corolla          4  gas  102466.0   
3965   16750        1985        chevrolet corvette          8  gas   24540.0   
20817  11886        2017            ford fusion se          4  gas   72000.0   
29377   9450        2014               kia sorento          4  gas   91776.0   
6114     500        2003              dodge dakota          6  gas  215457.0   
9607    4999        2012             nissan sentra          4  gas  125000.0   

      transmission       type paint_color  is_4wd date_posted  days_listed  \
41940    automatic      sedan      silver       0  2018-09-28            0   
43590    automatic      coupe      silver       0  2018-10-23            0   
35605    automatic      sedan         red       0  2018-08-02            0   
26876    automatic      truck      custom       1  2019-02-09            0   
23195    automatic      sedan      silver       0  2019-03-09            0   
3965     automatic  hatchback       white       0  2018-10-14            0   
20817    automatic      sedan     Unknown       0  2018-10-16            0   
29377    automatic        SUV      silver       1  2018-08-27            0   
6114     automatic     pickup     Unknown       0  2018-10-31            0   
9607     automatic      sedan      silver       0  2019-03-26            0   

       condition  
41940    salvage  
43590    salvage  
35605    salvage  
26876       fair  
23195       fair  
3965   excellent  
20817       good  
29377       good  
6114     salvage  
9607        fair